src/extractor/git.rs

```diff
@@ -0,0 +1,210 @@
+}
+
+impl Object {
+    fn new() -> Object {
+        Object {
+            ..Default::default()
+        }
+    }
+}
+
+pub fn new(repo: &repo::Repo) -> Result<Git, String> {
+    let git_data = match extrat_git_objects(&repo) {
+        Ok(d) => d,
+        Err(err) => return Err(format!("failed to extract git objects: {}", err).to_string()),
+    };
+
+    Ok(git_data)
+}
+
+pub fn extrat_git_objects(repo: &repo::Repo) -> Result<Git, git2::Error> {
+    let r = &repo.repo;
+    let mut walk = r.revwalk()?;
+
+    //Get all repository references
+    let mut oid: git2::Oid;
+    for rf in repo.repo.references()? {
+        //add each reference objects in the walker
+        if let Some(ref_name) = rf.unwrap().name() {
+            oid = r.revparse_single(ref_name)?.id();
+            walk.push(oid)?;
+        }
+    }
+
+    let mut objects: HashMap<String, Object> = HashMap::new();
+    //For each commit found from the references
+    for w in walk {
+        let mut obj = Object::new();
+        let oid = w?.clone();
+        let commit = r.find_commit(oid)?;
+
+        obj.kind = ObjectKind::Commit;
+
+        //Get commits
+        obj.commit = Some(Commit {
+            author: commit.author().to_string(),
+            sha: commit.id().to_string(),
+            message: commit.message().unwrap_or("").to_string(),
+            tree: commit.tree()?.id().to_string(),
+            committer: commit.committer().to_string(),
+            parents: (|| -> Vec<String> {
+                let mut ids = vec![];
+                for id in commit.parent_ids() {
+                    ids.push(id.to_string());
+                }
+                ids
+            })(),
+        });
+        //Add the commit object in the objects HashMap
+        objects.insert(oid.to_string(), obj);
+
+        //Add every git objects found during the tree object traversal
+        add_tree_objects(&commit.tree()?, &mut objects, &r)?;
+    }
+
+    Ok(Git {
+        objects,
+        ..Default::default()
+    })
+}
+
+fn add_tree_objects(
+    tree: &git2::Tree,
+    objects: &mut HashMap<String, Object>,
+    repo: &git2::Repository,
+) -> Result<(), git2::Error> {
+    tree.walk(git2::TreeWalkMode::PreOrder, |_, entry| {
+        let mut obj = Object::new();
+
+        if let Some(kind) = entry.kind() {
+            match kind {
+                //Create and add Tree objects
+                git2::ObjectType::Tree => {
+                    obj.kind = ObjectKind::Tree;
+                    obj.tree = Some(Tree {
+                        name: entry.name().unwrap_or("").to_string(),
+                        sha: entry.id().to_string(),
+                        filemode: entry.filemode(),
+                        objects: (|| -> Vec<String> {
+                            let mut objs = vec![];
+                            let t = repo.find_tree(entry.id()).unwrap();
+
+                            //We walk down the tree to find every blob objects and add them
```

Comment:
If I understand correctly we also find every tree objects and add them to our list of objects (even if we don't walk down those)